NAT source port metrics & table

tommyp1ckles opened the pull request:
Adds nat/stats which adds a module that computes stats for nat maps.
This uses the batched iteration of the nat map added in previous commits to count the number of allocated ports there are in each nat connection tuple: {source_ip, endpoint_ip, endpoint_port}.

This is necessary to monitor saturation of endpoint connections based on datapath allocation of source ports.

These are exported as the metric: `nat_endpoint_max_connection`

The "nat-stats" table in statedb, accessible via `cilium-dbg statedb dump`, ex:

```
root@kind-worker:/home/cilium# cilium statedb nat-stats
IPFamily   Proto    EgressIP                RemoteAddr                   Count
ipv4       TCP      10.244.1.89             10.244.3.86:4240             1
ipv4       TCP      10.244.1.89             10.244.0.170:4240            1
ipv4       TCP      172.18.0.2              172.18.0.5:4240              1
ipv4       TCP      172.18.0.2              172.18.0.3:4240              1
ipv4       TCP      172.18.0.2              172.18.0.5:6443              4
ipv4       ICMP     172.18.0.2              172.18.0.5                   6
ipv4       ICMP     172.18.0.2              172.18.0.3                   6
ipv6       TCP      [fc00:c111::2]          [fc00:c111::5]:4240          1
ipv6       TCP      [fd00:10:244:1::e8ce]   [fd00:10:244:3::6860]:4240   1
ipv6       TCP      [fc00:c111::2]          [fc00:c111::3]:4240          1
ipv6       TCP      [fd00:10:244:1::e8ce]   [fd00:10:244::b991]:4240     1
ipv6       ICMPv6   [fc00:c111::2]          [fc00:c111::5]               6
ipv6       ICMPv6   [fc00:c111::2]          [fc00:c111::3]               6
```

tommyp1ckles committed:
nat/stats: adds metric and table for nat table stats.

Adds nat/stats which adds a module that computes stats for nat maps.
This uses the batched iteration of the nat map added in previous
commits to count the number of allocated ports there are in each
nat connection tuple: {source_ip, endpoint_ip, endpoint_port}.

This is necessary to monitor saturation of endpoint connections
based on datapath allocation of source ports.

ex. for a node with NAT mappings for endpoint addr: 10.0.0.123:8080
each new connection will be mapped a source port creating a unique
tuple of {source_ip, allocated_port, remote_addr, remote_port}.
This means that we can only allocate within the ephemeral port range
within that "bucket".

nat/stats uses batched iteration to efficiently count these buckets
and use that to:

i) Store the top-k (defautl 32) bucket sizes in a statedb table.
ii) emit the max bucket size (for each ip family) as a metric.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
daemon: newDaemon returns promise for *option.DaemonConfig.

This resolves in the same way that the daemon promise resolves, but
resolves to *option.DaemonConfig notifying that newDaemon init tasks
have completed. Providing daemon config as a resolvable promise
allows code that is being added to the Daemon cells to wait for this
resolution without introducing cyclic package dependencies.

This is necessary as newDaemon may override fields in DaemonConfig such
as when doing kube proxy replacement init or device detection.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
cilium-dbg: add nat-stats table to statedb dump.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
pkg/maps/nat: expose global nat maps to hive.

Previously, global nat ip4/ip6 maps are accessed from the package variables.
To make accessing the maps easier for subsequent functionality this commit
adds pkg/maps/nat/cell.go which exports a cell.Module(...) of global maps via
the interfaces NatMap4 & NatMap6.

Because setup of nat maps depends on the outcome of EnableNodePort which
is not finalized until after newDaemon, this provides the maps via
promises.

Provided NAT maps are not used by loader, so we don't need to wrap with
MapOut[T] as that creates a dependency cycle in hive in newDaemonPromise.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
pkg/maps/nat: add batched stats counting of nat map.

This will be used to efficiently walk the map to generate stats
for nat table port utilization.

Lookup batch on LRU hash map may fail if the buffer passed is not big enough to
accomadate the largest bucket size in the LRU map [1]

This sets a dynamic buffer size based on the expected bucket size within
the LRU map.

Because bucket size, cannot be known, we take a probabilistic approach by starting
with a size that is over sqrt(max_entries * 2) which is the number of elements
before we expect to see a collision (i.e. elements hashed into the same bucket).
For cases where we hit this case, given this assumed bucket size, we will double the size of the
chunks and try again, up to 3 times, leading to a 3 fold increase.

Default NAT map size is 262144 -> 2^ceil(log2(sqrt(262144 * 2))) = 1024, with key + entry size
being ~ 432 bits, this means we'll need to allocate 55kb to accommodate this iteration.

To avoid unbounded growth, each ENOSPC will result in a doubling of the chuck chunkSize
which will persist into subsequent calls of Stats, up to a maximum of 3 (fold-increase).

[1] https:elixir.bootlin.com/linux/latest/source/kernel/bpf/hashtab.c#L1776

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

christarazi commented on pkg/maps/nat/nat.go:
```diff
@@ -98,6 +102,73 @@ func NewMap(name string, family IPFamily, entries int) *Map {
+}
+
+func applyBatch[TupleType any, EntryType any](m *Map, fn func([]TupleType, []EntryType, int), chunkSize uint32) (count int, err error) {
+	kout := make([]TupleType, chunkSize)
+	vout := make([]EntryType, chunkSize)
+
+	var cursor ebpf.MapBatchCursor
+	for {
+		c, batchErr := m.BatchLookup(&cursor, kout, vout, nil)
+		count += c
+		fn(kout, vout, c)
```
If on the last chunk with size less than `chunkSize`, will `kout` and `vout` contain old elements from the previous chunk? If so, this would affect the logic invoked in `fn()` right?

tommyp1ckles replied:
If this fails due to the ENOSPC error, then we just start over with a whole new `applyBatch()` from the beginning, so there shouldn't be any issues with data integrity.

**note**: in testing, I was having a very hard time actually reproducing this scenario so I don't it will happen very often. The starting batch size function seems to do a good job of coming up with a good size to use based on map size.

christarazi replied:
I meant stale, old entries from the previous chunk left over. If that happened, that would cause the wrong result to be returned, right?

tommyp1ckles replied:
I might be misunderstanding, the map cursor is initialized in the scope of `applyBatch(...)` so if this fails due to ENOSPC, this function returns with that error and `applyBatchReliably(...)`will double the chunkSize and start over with a cursor value of 0.

christarazi replied:
Ah gotcha.

christarazi commented on pkg/maps/nat/stats/cell.go:
```diff
@@ -0,0 +1,95 @@
+	),
+	cell.Config(Config{
+		NATMapStatInterval:       30 * time.Second,
+		NatMapStatKStoredEntries: 32,
```
How did you decide on these defaults?

tommyp1ckles replied:
For NatMapStatKStoredEntries: this is just how many of the top-k entries we're planning on storing in statedb (and eventually an optional metric) by default. 32 seemed reasonable.

NATMapStatInterval was also just based on a reasonable interval, however I should expand on this and do some benchmarking based on this interval.

christarazi replied:
It would be good to leave either a code comment or reference it in the commit msg just to have a record.

tommyp1ckles replied:
To follow up here, here's some basic testing that these settings are and the overall change set don't cause any apparant memory regressions in a "general" case (in this test case I ran EGW while forwarding 40k connections with my changes and without).

![control](https://github.com/cilium/cilium/assets/5881740/daafe818-d02e-4071-8a3a-a41ced9de1a6)
![test](https://github.com/cilium/cilium/assets/5881740/7dc69f66-7b2e-4bd8-888e-3fa82d333261)

tommyp1ckles replied:
I'll add some comments explaining this context

christarazi commented on pkg/maps/nat/stats/cell.go:
```diff
@@ -0,0 +1,95 @@
+func (def Config) Flags(flags *pflag.FlagSet) {
+	flags.Duration("nat-map-stats-interval", def.NATMapStatInterval, "Interval upon which nat maps are iterated for stats")
+	flags.Int("nat-map-stats-entries", def.NatMapStatKStoredEntries, "Number k top stats entries to store locally in statedb")
+	flags.MarkHidden("nat-map-stats-entries")
```
I see `nat-map-stats-entries` marked as hidden. Any reason not to also mark `nat-map-stats-interval` as hidden?

tommyp1ckles replied:
Previously, nat-map-stats-entries could've had a big memory/perf usage impact, but now the impact would be minimal so I think its best to just remove the MarkHidden.

christarazi commented on pkg/maps/nat/stats/cell.go:
```diff
@@ -0,0 +1,95 @@
+		Name: "byTuple",
+		FromObject: func(s NatMapStats) index.KeySet {
+			return index.StringSlice(
+				[]string{fmt.Sprintf("%s/%s %s %s:%d", s.Type, s.Proto, s.EgressIP, s.EndpointIP, s.RemotePort)})
```
I'm not sure if this is a "hotpath" function, but `fmt.Sprintf` can be quite inefficient and expensive. This seems like a straightforward string concat, so if it's hotpath function, probably worth using `+` instead.

tommyp1ckles replied:
good point! thanks

christarazi commented on pkg/maps/nat/stats/stats.go:
```diff
@@ -0,0 +1,291 @@
+}
+
+func (m *Stats) run(ctx context.Context, hr cell.Health) {
+	controller.NewManager().UpdateController("nat-stats", controller.ControllerParams{
```
Any reason not to use the timer jobs? We seem to be preferring those over controllers, at the least for new code.

tommyp1ckles replied:
Switched to using jobs, much cleaner now

joamaki replied:
This seems to still be here. Didn't push yet?

christarazi commented on pkg/maps/nat/stats/stats.go:
```diff
@@ -0,0 +1,291 @@
+			}
+		})
+		if err != nil {
+			m.logger.WithError(err).Error("failed to count ipv4 nat map entries")
```
This will be a user-facing error log msg, so I think it should provide a bit more context as to the consequences of erroring here, and what might the user want to do or what Cilium will do about it.

christarazi replied:
Same applies to other log msgs with WARN level and above.

tommyp1ckles replied:
good point

christarazi commented on pkg/maps/nat/stats/stats.go:
```diff
@@ -0,0 +1,291 @@
+					tupleToPortCount[key] = ports
+				}
+			}
+		})
```
Seems like this code could be DRY'd up along with its IPv6 variant

tommyp1ckles replied:
Yeah it does 😅  but when I was looking at DRYing this up I came to the conclusion that trying to make this more abstract might make the code overall less readable since we'd have to implement and manipulate the NAT keys behind interfaces w/ getters/setters... ex something like:

```go
func countEntry[T KeyInterface, K EntryInterface](...) {
    key.SetSourcePort(swapBytes(key.GetSourcePort())
    if key.GetFlags() == tuple.TUPLE_F_IN && key.GetNextHeader() == u8proto.TCP  || 
        key.GetNextHeader() == familyICMPType {
    // and so on ... 
```

but I'm open to suggestions 😄

christarazi reviewed (commented):
Nice work! I took a first-pass looking at the core implementation and left a couple of comments.

ysksuzuki commented on pkg/maps/nat/stats/stats.go:
```diff
@@ -0,0 +1,320 @@
+			natMapStatsEntriesName, minNatMapStatKStoredEntries, maxNatMapStatKStoredEntries)
+	}
+
+	maxAvailPorts := 65535 - (option.Config.NodePortMax + 1)
```
Does this mean `NODEPORT_PORT_MAX_NAT` - `NODEPORT_PORT_MIN_NAT`?
https://github.com/cilium/cilium/blob/1c9d968f832659452977cede8b86db68a6697bb4/pkg/datapath/linux/config/config.go#L546

Can we define a constraint for NODEPORT_PORT_MAX_NAT somewhere and comment that it corresponds to  `NODEPORT_PORT_MAX_NAT` - `NODEPORT_PORT_MIN_NAT`?

ysksuzuki commented on pkg/maps/nat/stats/stats.go:
```diff
@@ -0,0 +1,320 @@
+				key := keys[i]
+				key.SourcePort = swapBytes(key.SourcePort)
+				if key.Flags == tuple.TUPLE_F_IN &&
+					(key.NextHeader == u8proto.TCP || key.NextHeader == u8proto.ICMP) {
```
Why is UDP excluded here?

tommyp1ckles replied:
That should be included, ive added that

ysksuzuki commented on pkg/maps/nat/stats/cell.go:
```diff
@@ -0,0 +1,101 @@
+
+func newMetrics() Metrics {
+	return Metrics{
+		LocalPorts: metric.NewGaugeVecWithLabels(metric.GaugeOpts{
```
Can we document `nat_endpoint_max_connection` here?
https://github.com/cilium/cilium/blob/579d3b624444792c8b65002c5c7f284f9d98ec90/Documentation/observability/metrics.rst

dylandreimerink commented on pkg/maps/nat/cell.go:
```diff
@@ -0,0 +1,100 @@
+				}
+				if cfg.EnableIPv6 {
+					res6.Resolve(ipv6Nat)
+				}
```
I am not sure if promises are the ideal way to do optional values. But if we assume they are, then I would also reject the promise in these cases where you are not going to resolve them, that will help the receiving side to not pointlessly wait for a ctx to expire.

tommyp1ckles replied:
Fixed this, me and @joamaki also [brought up concerns](https://github.com/cilium/cilium/pull/32152#discussion_r1603131450) with using promises here.  I've created an issue to track this problem more broadly: https://github.com/cilium/cilium/issues/32557.

dylandreimerink reviewed (changes requested):
The inline comment I left is the only blocker for me.

I understand the reasoning behind the daemon config promise, but I feel like we should get rid of this mutating config object instead of using a promise to delay certain action until it has stabilized. I understand that can't change in one PR,  so will not block on this, but would like to express that.

Same goes for the LRU bulk sizing and backoff. I suspect the issues you are dealing with is the same as in https://github.com/cilium/ebpf/issues/1390. This honestly feels like something we should address at the library level since it will be a more common issue. So I would suggest discussing upstreaming this in cilium/ebpf. Again, something of after this PR.

tommyp1ckles committed:
datapath/config: add constant for NODEPORT_PORT_MAX_NAT.

This will be useful for counting stats for nat map utilization
in subsequent commits.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
documentation: add section for NAT metrics

Documents nat_endpoint_max_connection as well as giving a brief
overview of NAT based metrics.

This is being put into a separate new section as there will likely
be more similar metrics added in the future.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

joamaki commented on pkg/maps/nat/stats/cell.go:
```diff
@@ -0,0 +1,102 @@
+		Name: "byTuple",
+		FromObject: func(s NatMapStats) index.KeySet {
+			return index.StringSlice([]string{
+				s.Type + "/" + s.Proto + " " + s.EgressIP + " " +
```
I have a few recommendations here:
- This allocates an unnecessary slice that gets immediately consumed. Use `index.NewKeySet` directly and construct the byte slice / string efficiently. E.g. use a string/byte builder.
- Composite keys like this shouldn't be using the string type directly. I would recommend making a `NatMapKey` struct containing the tuple and then have a `Key() index.Key` function that you'd use in `FromObject` and `FromKey`. This way you don't have to ponder how to correctly construct the key when doing a query.

tommyp1ckles replied:
Updated it to use a Key function on NatMapStat, lmk if that's what you meant.

joamaki commented on pkg/maps/nat/stats/cell.go:
```diff
@@ -0,0 +1,102 @@
+	natMapStatsEntriesName      = "nat-map-stats-entries"
+)
+
+var Cell = cell.Module(
```
Add a comment describing what this cell provides to the application.

joamaki replied:
Really nice!

joamaki commented on pkg/maps/nat/cell.go:
```diff
@@ -0,0 +1,104 @@
+
+var MapDisabled = fmt.Errorf("nat map is disabled")
+
+var Cell = cell.Module(
```
Add a comment describing what this cell provides to the application.

joamaki commented on pkg/maps/nat/cell.go:
```diff
@@ -0,0 +1,104 @@
+	"nat-maps",
+	"NAT Maps",
+	cell.Provide(func(lc cell.Lifecycle, cfgPromise promise.Promise[*option.DaemonConfig]) (promise.Promise[NatMap4], promise.Promise[NatMap6]) {
+		var ipv4Nat, ipv6Nat *Map
```
Use of promises isn't great, but it's difficult to get out of this without pulling all the relevant fields out from DaemonConfig into separate config structs, and then having a validation/probing that takes the "user config" and produces the "derived config". Though when probing is involved our options currently are to either make the config only "usable" from start hooks on-wards (constructors would see unfilled config), or use a promise which makes it more explicit (though using the promise from constructor would just hang).

I think you also raised a 3rd option that would be add a "3rd stage" to hive: 1) populate to grab config flags, 2) inject parsed flags, validate and probe configs, 3) run all invokes and construct the application. We could then make the "probing" no-op for graph inspection. I don't like the complexity of this though. On the other hand if we don't have this, then we have the problem of how do we deal with "disabled" things that we only know are disabled when starting. Which then requires e.g. constructing promises which are not friendly to use.

So I think this is OKish to do Promise[DaemonConfig] for short term, but hopefully this doesn't stick around too long. And I don't really have a good solution around this yet.

tommyp1ckles replied:
> So I think this is OKish to do Promise[DaemonConfig] for short term, but hopefully this doesn't stick around too long. And I don't really have a good solution around this yet.

Yeah, I ended up here after a few different approaches that where all [even less satisfying](https://github.com/cilium/cilium/pull/32275/files#diff-c0e8dc92b48dcab08b42d07a5223eb03c267c8484f975a66f423ae85d9cc6ea2R1657).

In the nat/stats modules use case specifically, the promise "chain" ends here, as it is designed to serve as a data source using a table type, so consumers/metrics just have to wait for everything to resolve before being updated which is fine.

But, for the nat map, I'd like to have this be known prior to runtime, but as you mentioned, there's no great way to do that right now.  This is likely a one-off situation though as most maps need to be exported using `bpf.MapOut[T]`which doesn't work due to dependency cycles.

tommyp1ckles replied:
@joamaki I've linked the issue addressing this problem in the todo comment, going to resolve as this seems like a bigger discussion related to how we break-down and refactor *DaemonConfig in the future [1]

[1] Related Issue: https://github.com/cilium/cilium/issues/32557

joamaki commented on pkg/maps/nat/stats/cell.go:
```diff
@@ -0,0 +1,102 @@
+	}
+)
+
+func newTables(db *statedb.DB) (statedb.RWTable[NatMapStats], error) {
```
Exporting this is often useful because if some new component starts using `Table[NatMapStats]`, then it can also construct the table directly in its tests without having to use this actual cell to gain access to it. Here of course this doesn't really apply as both this and the table itself are completely private to this module, but though I'd mention it as a general guideline (which reminds me I really need to create a list of guidelines...).

Though would there be any harm in exporting `Table[NatMapStats]` (not RWTable)? Would this data be useful in any other part of our apps?

tommyp1ckles replied:
Actually, this is intended to be exported as it will be used as a data source in further work so i'll add the `statedb.RWTable[NatMapStats].ToTable`to rexport the read only table.

joamaki commented on pkg/maps/nat/stats/stats.go:
```diff
@@ -0,0 +1,325 @@
+		_, err := m.natMap6.ApplyBatch6(func(keys []tuple.TupleKey6, vals []nat.NatEntry6, size int) {
+			for i := 0; i < size; i++ {
+				key := keys[i]
+				key.SourcePort = swapBytes(key.SourcePort)
```
Not knowing the details, but this seems wrong. Do we know for a fact the source and target endianess? Shouldn't this use `binary.BigEndian.PutUint16` and friends?

tommyp1ckles replied:
These are inserting by the datapath directly from packet ctx, thus they are in network order.  However I should be using the `tuple.MapKey.(ToHost)` func to make this clear and for code reuse 👍

julianwiedmann commented on pkg/maps/nat/stats/stats_test.go:
```diff
@@ -0,0 +1,132 @@
+			mapKey.TupleKey4.DestPort = 9000 + i
+			mapKey.Flags = tuple.TUPLE_F_IN
+			mapKey.NextHeader = u8proto.TCP
+
```
Could you spell `mapKey.TupleKey4.SourcePort = 0` out here (and below)? Or even use some other constant port number? It took me too long to grasp what's going on :grin:.

We could also add some identical UDP entries here (and/or vary the `SourcePort`). To emphasize that the statistics are truly sensitive to **any** difference in the `(SourceIP, SourcePort, DstIP, L4-Proto)` tuple, and we don't count false-positives.

tommyp1ckles replied:
Good idea, I'll set them to mirror the endpoint IP somewhat by making them 8000 + <final_ip_octet>... ex:

```
10.0.0.1:8001 
10.0.0.1:8002
...
```

tommyp1ckles committed:
nat/stats/cell: export statedb.Table[NatMapStats] to hive.

This will serve as the primary interface for future components
to access nat map stats data.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

qmonnet commented on Documentation/observability/metrics.rst:
```diff
@@ -1271,3 +1271,21 @@ which is passed as the ``controller-group-metrics`` configuration
 for all controller groups. The special name "none" is also supported.
+
+NAT
+~~~~~~~~~~~
```
```suggestion
NAT
~~~
```

qmonnet commented on pkg/maps/nat/stats/stats.go:
```diff
@@ -0,0 +1,323 @@
+
+	if params.Config.NatMapStatKStoredEntries > maxNatMapStatKStoredEntries ||
+		params.Config.NatMapStatKStoredEntries < minNatMapStatKStoredEntries {
+		return nil, fmt.Errorf("nat-stats config: %q must be betweed [%d, %d]",
```
```suggestion
		return nil, fmt.Errorf("nat-stats config: %q must be between [%d, %d]",
```

qmonnet commented on pkg/maps/nat/nat_batch_test.go:
```diff
@@ -0,0 +1,43 @@
+
+func TestApplyPatch(t *testing.T) {
+	testutils.PrivilegedTest(t)
+	m := NewMap("test_snat_map", IPv4, 1<<18) // aprox default map size.
```
```suggestion
	m := NewMap("test_snat_map", IPv4, 1<<18) // approximative default map size.
```

qmonnet commented on Documentation/observability/metrics.rst:
```diff
@@ -1271,3 +1271,21 @@ which is passed as the ``controller-group-metrics`` configuration
+======================================== ================================================== ========== ========================================================
+
+These metrics are for monitoring Cilium NAT mapping function as is used for
+egress-gateway and bpf-masquerade features.
```
Really a suggestion, keep your version if you prefer.

```suggestion
These metrics are for monitoring Cilium's NAT mapping function.
NAT mapping is used by features such as Egress Gateway and BPF masquerading.
```

tommyp1ckles replied:
Agreed, yours is less clunky 👍

qmonnet commented on Documentation/observability/metrics.rst:
```diff
@@ -1271,3 +1271,21 @@ which is passed as the ``controller-group-metrics`` configuration
+for any connection 3-tuple of egress-IP, endpoint-IP and endpoint-port in Cilium NAT maps.
+This is useful for monitoring how close a Node is to exhausting the max number of connections
+to any single remote endpoint.
+
```
```suggestion
The ``nat_endpoint_max_connection`` metric reports the maximum proportion of unique source ports allocated
for any connection 3-tuple of egress-IP, endpoint-IP and endpoint-port in Cilium NAT maps.
This is useful for monitoring how close a Node is to exhausting the maximum number of connections
to any single remote endpoint.
```

Note: the first sentence is not super clear in my opinion. “The max proportion of unique source ports allocated [...]” sounds like this is a value computed for a theoretical maximum (which I guess would be 100%), rather than a value that reflects the current status of the table. Not sure how to improve it. Maybe “highest” rather than “max”? Or “current maximum”? :thinking:. It would also be useful to make it very explicit what the “proportion” is here (number of unique source ports, out of what exactly?).

qmonnet commented on pkg/maps/nat/cell.go:
```diff
@@ -0,0 +1,115 @@
+				// TODO: Refactor ctmap gc Enable() such that it can use the map descriptors from
+				// here so we can move all nat map creation logic into here.
+				if err := ipv4Nat.Open(); err != nil {
+					return fmt.Errorf("open ipv4 nat map: %w", err)
```
Please try to use the right capitalisation in user-facing messages, here and elsewhere in the code. Ideally I'd recommend doing the same for comments in the code as well, but I don't mind too much.

```suggestion
					return fmt.Errorf("open IPv4 NAT map: %w", err)
```

qmonnet commented on Documentation/observability/metrics.rst:
```diff
@@ -1271,3 +1271,21 @@ which is passed as the ``controller-group-metrics`` configuration
+======================================== ================================================== ========== ========================================================
+Name                                     Labels                                             Default    Description
+======================================== ================================================== ========== ========================================================
+``nat_endpoint_max_connection``          ``family``                                         Enabled    Max utilization of source ports on a egress-ip/external endpoint tuple, as a proportion of max nat per-connection ports.
```
Some more suggestions:

```suggestion
``nat_endpoint_max_connection``          ``family``                                         Enabled    Maximum utilization of source ports on an {egress-ip, external endpoint} tuple, as a proportion of maximum NAT per-connection ports
```

Same as the other comment, I think the description could be improved. The double meaning for “max“ (as in, current highest value vs. maximum value allowed by the Agent) is slightly confusing.

tommyp1ckles replied:
👍  Makes sense,  I'll rewrite it to try to make it more understandable for some who isn't familiar with NAT/EGW.

tommyp1ckles replied:
@qmonnet I've rewritten this (and the description below) to be easier to understand, let me know what you think 😄 .

qmonnet replied:
It does read better, thank you!

qmonnet reviewed (changes requested):
Nice work!

I've got some minor comments regarding the docs (or some strings in the code), but it should be easy to address, please see inline below. As you've certainly noticed, you've got a number of CI failure - with my docs-structure hat on, I'll say, “please don't forget to update the reference docs” :slightly_smiling_face:.

ldelossa commented on pkg/maps/nat/nat.go:
```diff
@@ -98,6 +102,73 @@ func NewMap(name string, family IPFamily, entries int) *Map {
 }
 
+func startingChunkSize(maxEntries int) int {
+	bucketSize := math.Sqrt(float64(maxEntries * 2))
```
Can you spell out a bit more about this math here? I'm curious about how we obtain this bucketSize.

tommyp1ckles replied:
I took this from a discussion that @antonipp and @christarazi had regarding a similar usage of batched iteration 😄.

I _think_ this is this approximation of the [Birthday Paradox ](https://en.wikipedia.org/wiki/Birthday_problem#Square_approximation) where we set P(n) = 1.

ldelossa reviewed (approved):
Overall, changes look good to me. A bit curious about some of the math used, but that's just for my own knowledge. Nice job.

ttarczynski commented on Documentation/observability/metrics.rst:
```diff
@@ -1271,3 +1271,24 @@ which is passed as the ``controller-group-metrics`` configuration
+======================================== ================================================== ========== ========================================================
+Name                                     Labels                                             Default    Description
+======================================== ================================================== ========== ========================================================
+``nat_endpoint_max_connection``          ``family``                                         Enabled    Saturation of the most the most saturated distinct NAT mapped connection, in terms of egress-IP and remote endpoint address.
```
Should be: 
```
Saturation of the most saturated distinct NAT mapped connection
```

Instead of:
```
Saturation of the most the most saturated distinct NAT mapped connection
```

tommyp1ckles replied:
Oh nice catch 😄  thanks

ttarczynski commented on Documentation/observability/metrics.rst:
```diff
@@ -1271,3 +1271,24 @@ which is passed as the ``controller-group-metrics`` configuration
+These metrics are for monitoring Cilium's NAT mapping functionality. NAT is used by features such as Egress Gateway and BPF masquerading.
+
+The NAT map holds mappings for masqueraded connections. Connection held in the NAT table that are masqueraded with the
+same egress-IP and are going to the same remote endpoint address all require a unique source port for the mapping.
```
Can we clarify what "remote endpoint" means? I assume it's typically IP (v4|v6) address + (TCP|UDP) port.

Maybe to add it here:

> Connection held in the NAT table that are masqueraded with the same egress-IP and are going to the same remote endpoint address all require a unique source port for the mapping.

Could we mention the destination port explicitly too?

> [...] going to the same remote endpoint (address and port) [...]

derailed reviewed (approved):
@tommyp1ckles Nice work Tom!

tommyp1ckles committed:
datapath: add fake nat.NatMap{4,6} provider.

These are rejected as 'MapDisabled', such that the nat/stats
is disabled.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

ysksuzuki reviewed (approved):
LGTM, Thanks!

tommyp1ckles committed:
datapath/probes: add probe function to check for batch lookup api.

Kernel >=5.6 includes batched lookup API, this provides a probe function
for detecting this functionality by create a small temporary map and
attempting to do a batch lookup syscall.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

julianwiedmann commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -300,6 +300,8 @@ cilium-agent [flags]
       --mtu int                                                   Overwrite auto-detected MTU of underlying network
       --multicast-enabled                                         Enables multicast in Cilium
+      --nat-map-stats-entries int                                 Number k top stats entries to store locally in statedb (default 32)
+      --nat-map-stats-interval duration                           Interval upon which nat maps are iterated for stats (default 30s)
```
Should we also say that `duration = 0` disables the stats collection? Or is that obvious?
